File: j-LIB-core/src/main/java/io/github/dot166/jlib/game/Obstacle.java
package io.github.dot166.jlib.game;

import static io.github.dot166.jlib.game.GameView.PARAMS;

import android.content.Context;
import android.graphics.Rect;
import android.view.View;

public class Obstacle extends View implements GameObject {
    public float h;

    public final Rect hitRect = new Rect();

    public Obstacle(Context context, float h) {
        super(context);
        setBackgroundColor(0xFFFF0000);
        this.h = h;
    }

    public boolean intersects(Player p) {
        final int N = p.corners.length / 2;
        for (int i = 0; i < N; i++) {
            final int x = (int) p.corners[i * 2];
            final int y = (int) p.corners[i * 2 + 1];
            if (hitRect.contains(x, y)) return true;
        }
        return false;
    }

    public boolean cleared(Player p) {
        final int N = p.corners.length / 2;
        for (int i = 0; i < N; i++) {
            final int x = (int) p.corners[i * 2];
            if (hitRect.right >= x) return false;
        }
        return true;
    }

    @Override
    public void step(long t_ms, long dt_ms, float t, float dt) {
        setTranslationX(getTranslationX() - PARAMS.TRANSLATION_PER_SEC * dt);
        getHitRect(hitRect);
    }
}

File: j-LIB-core/src/main/java/io/github/dot166/jlib/game/Obstacle.kt
package io.github.dot166.jlib.game

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Rect
import android.view.View

@SuppressLint("ViewConstructor")
open class Obstacle(context: Context?, h: Float) : View(context), GameObject {
    @JvmField
    var h: Float

    @JvmField
    val hitRect: Rect = Rect()

    init {
        setBackgroundColor(-0x10000)
        this.h = h
    }

    fun intersects(p: Player): Boolean {
        val n = p.corners.size / 2
        for (i in 0..<n) {
            val x = p.corners[i * 2].toInt()
            val y = p.corners[i * 2 + 1].toInt()
            if (hitRect.contains(x, y)) return true
        }
        return false
    }

    fun cleared(p: Player): Boolean {
        val n = p.corners.size / 2
        for (i in 0..<n) {
            val x = p.corners[i * 2].toInt()
            if (hitRect.right >= x) return false
        }
        return true
    }

    override fun step(tMs: Long, dtMs: Long, t: Float, dt: Float) {
        translationX = translationX - GameView.PARAMS.TRANSLATION_PER_SEC * dt
        getHitRect(hitRect)
    }
}
